src/main/java/com/aws/greengrass/certificatemanager/certificate/CISShadowMonitor.java

```diff
@@ -259,53 +464,54 @@ private synchronized void processCISShadow(int version, Map<String, Object> desi
-        LOGGER.info("Publishing to get shadow topic");
-        GetShadowRequest getShadowRequest = new GetShadowRequest();
-        getShadowRequest.thingName = shadowName;
-        iotShadowClient.PublishGetShadow(getShadowRequest, QualityOfService.AT_LEAST_ONCE)
-                .exceptionally(e -> {
-                    LOGGER.atWarn().cause(e).log("Unable to retrieve CIS shadow");
-                    return null;
-                });
-    }
+        @SuppressWarnings("PMD.SignatureDeclareThrowsException")
+        private List<String> fetchHostAddressesFromCIS() throws Exception {
+            RetryUtils.runWithRetry(
+                    GET_CONNECTIVITY_RETRY_CONFIG,
+                    connectivityInfoProvider::getConnectivityInfo,
+                    "get-connectivity",
+                    LOGGER
+            );
+            return connectivityInfoProvider.getCachedHostAddresses();
+        }
 
-    private void unsubscribeFromShadowTopics() {
-        if (connection != null) {
-            LOGGER.atDebug().log("Unsubscribing from CIS shadow topics");
-            String topic = String.format(SHADOW_UPDATE_DELTA_TOPIC, shadowName);
-            connection.unsubscribe(topic);
+        /**
+         * Asynchronously update the CIS shadow's <b>reported</b> state for the given shadow version.
+         *
+         * @param version CIS shadow version
+         * @param desired CIS shadow <b>desired</b> state
+         */
+        private void updateCISShadowReportedState(int version, Map<String, Object> desired) {
+            LOGGER.atInfo().kv(VERSION, version).log("Reporting CIS version");
+            UpdateShadowRequest updateShadowRequest = new UpdateShadowRequest();
+            updateShadowRequest.thingName = shadowName;
+            updateShadowRequest.version = version;
+            updateShadowRequest.state = new ShadowState();
+            updateShadowRequest.state.reported = desired == null ? null : new HashMap<>(desired);
+            iotShadowClient.PublishUpdateShadow(updateShadowRequest, QualityOfService.AT_LEAST_ONCE)
+                    .exceptionally(e -> {
+                        LOGGER.atWarn().kv(VERSION, version).cause(e).log("Unable to report CIS shadow version");
+                        return null;
+                    });
+        }
 
-            topic = String.format(SHADOW_GET_ACCEPTED_TOPIC, shadowName);
-            connection.unsubscribe(topic);
+        /**
+         * Signal that current work has been completed.
+         * Intended for unit testing.
+         */
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        private void notifyWorkComplete() {
+            if (onShadowProcessingWorkComplete != null) {
+                try {
+                    onShadowProcessingWorkComplete.run();
+                } catch (Exception e) {
+                    LOGGER.atDebug().cause(e).log("Unable to notify work complete");
+                }
+            }
```

Comment:
This exists solely for tests, correct? Could you remove this if you used a fake cert generator instead?